### PR TITLE
improve the async process in collector

### PR DESCRIPTION
**Pull request by yangli2014**

### CODE REVIEW

**Comment by yangli2014** [comment]
This new version seems to have some issues when testing in karaf. It takes much longer than before.

**Comment by yangli2014** [comment]
The thread and pool size have to be configured big enough when collecting all the keys for a node.

File: plugin/src/test/java/org/opennms/plugins/zabbix/ZabbixAgentCollectorTest.java [hunk 1]
```diff
@@ -25,8 +26,8 @@
 
     @Rule
     public MockZabbixAgent zabbixAgent = new MockZabbixAgent();
-    private int threadSize = 4;
-    private int poolSize = 10;
+    private int threadSize = 10;
```
Comment (j-white) [inline plugin/src/test/java/org/opennms/plugins/zabbix/ZabbixAgentCollectorTest.java]:
```suggestion
Can we create a default constructor on the ZabbixClientFactory and avoid having these values in the test?
```

File: plugin/src/test/java/org/opennms/plugins/zabbix/ZabbixAgentClientTest.java [hunk 2]
```diff
@@ -14,30 +15,41 @@
...
 
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+
 public class ZabbixAgentClientTest {
 
     @Rule
     public MockZabbixAgent zabbixAgent = new MockZabbixAgent();
-    private static int threadSize = 4;
+    private EventLoopGroup group = new NioEventLoopGroup(4);
```
Comment (j-white) [inline plugin/src/test/java/org/opennms/plugins/zabbix/ZabbixAgentClientTest.java]: Would be nicer to use the factory, and avoid creating the group in here.

File: plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java [hunk 3]
```diff
@@ -67,47 +69,34 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
                     .setStatus(CollectionSet.Status.SUCCEEDED);
 
             // Process the template
+            List<CompletableFuture<Void>> futureList = new ArrayList<>();
```
Comment (j-white) [inline plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: Unused?

File: plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java [hunk 4]
```diff
@@ -67,47 +69,34 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
...
+                                    }));
+                        } catch (ZabbixNotSupportedException | IOException e) {
+                            // pass
                         }
-                    } catch (ZabbixNotSupportedException e) {
-                        // pass
                     }
-                }
+                    try {
+                        CompletableFuture.allOf(ruleFutureList.toArray(new CompletableFuture[ruleFutureList.size()])).get();
```
Comment (j-white) [inline plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: We shouldn't block here - when these futures are complete, we should be able to complete the future that is returned by the function.

File: plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java [hunk 5]
```diff
@@ -67,47 +69,34 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
...
-                                String value = client.retrieveData(effectiveKey).get();
-                                addValueToMapper(rule, item, value, resourceBuilder);
-                                didAddAttribute = true;
-                            }
-
-                            // Only add the resource if we found 1+ items
-                            if (didAddAttribute) {
-                                collectionSetBuilder.addCollectionSetResource(resourceBuilder.build());
-                            }
+                processItems(client, null, template.getItems(), nodeResourceBuilder).thenRunAsync(() -> {
```
Comment (j-white) [inline plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: We should be able to process the items, and the discovery rules in parallel - no need to chain these or run them in sequence.

File: plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java [hunk 6]
```diff
@@ -119,6 +108,39 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
...
+            // Create a new resource for the entry
+            final ImmutableCollectionSetResource.Builder<GenericTypeResource> resourceBuilder = ImmutableCollectionSetResource.newBuilder(GenericTypeResource.class)
+                    .setResource(ImmutableGenericTypeResource.newBuilder()
+                            .setNodeResource(nodeResource)
+                            .setType(resourceType.getName())
+                            .setInstance(resourceTypeGenerator.getIndex(rule, entry))
+                            .build());
+
+            // Process the items in the rule
+            boolean didAddAttribute = false;
```
Comment (j-white) [inline plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: Unused, can be removed.

File: plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java (hunk 5 above)
Comment (yangli2014) [reply plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: Good idea

File: plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java [hunk 7]
```diff
@@ -65,48 +64,28 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
...
-                            // Only add the resource if we found 1+ items
-                            if (didAddAttribute) {
-                                collectionSetBuilder.addCollectionSetResource(resourceBuilder.build());
-                            }
-                        }
-                    } catch (ZabbixNotSupportedException e) {
+                        templateFutures.add(client.discoverData(rule.getKey())
+                                .thenAccept(list -> {
+                                    try {
+                                        processEntries(client, rule, list, collectionSetBuilder, nodeResource, resourceTypeGenerator);
```
Comment (j-white) [inline plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: The future returned by `processEntries` are not used. I would expect them to be part of the chain.

File: plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java [hunk 8]
```diff
@@ -65,48 +64,28 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
...
+                                    } catch (IOException e) {
```
Comment (j-white) [inline plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: IOException shouldn't be thrown by this - should return a future with exception instead.

Comment (yangli2014) [reply plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java]: Good catch.

**Review by j-white — APPROVED** [review]
Looks good. Thanks for all the updates!

**Pull request merged**